core/src/main/java/io/crate/types/TableType.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.types;
+
+import io.crate.Streamer;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * A type which represents a table retrieved by a subquery.
+ * May be used in an IN or ANY operation, e.g. select 1 where 1 = ANY(Select 1)
+ */
+public class TableType extends DataType<Collection> implements CollectionType, Streamer<Object> {
+
+    public static final int ID = 102;
+    private DataType<?> valueType;
```

Comment:
consider calling this rowType, but then...

actually this type is not a generic TableType it is a SingleColumn table type. it does not work with multiple columns. also it is not future prove since let's say we want to use a special complex type as type of the row in the future, the semantic is different because the valueType here is actually the valueType of the first column of each row